src/tools/schema2graph.ts

```diff
@@ -40,9 +40,11 @@ export class SchemaSource {
 
 export class SchemaNode {
   constructor(
-    readonly schema: Schema,
+    readonly schema: Schema | null,
     readonly particleSpec: ParticleSpec,
-    readonly allSchemaNodes: SchemaNode[]
+    readonly allSchemaNodes: SchemaNode[],
+    // Type variable associated with the schema node.
+    readonly fromVariable: string = ''
```

Comment:
type variable can have an empty name e.g. `data: reads ~`
So I would recommend representing lack of type variable as undefined or null